server/src/main/java/org/elasticsearch/index/IndexMode.java

```diff
@@ -43,6 +51,27 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
         private String error(Setting<?> unsupported) {
             return "[" + IndexSettings.MODE.getKey() + "=time_series] is incompatible with [" + unsupported.getKey() + "]";
         }
+
+        @Override
+        public void completeMappings(MappingParserContext context, RootObjectMapper.Builder builder) {
+            Optional<Mapper.Builder> timestamp = builder.getBuilder("@timestamp");
+            if (timestamp.isEmpty()) {
+                builder.add(
+                    new DateFieldMapper.Builder(
+                        "@timestamp",
+                        DateFieldMapper.Resolution.MILLISECONDS,
+                        DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER,
+                        context.scriptCompiler(),
+                        DateFieldMapper.IGNORE_MALFORMED_SETTING.get(context.getSettings()),
+                        context.getIndexSettings().getIndexVersionCreated()
+                    )
+                );
+                return;
+            }
+            if (false == timestamp.get() instanceof DateFieldMapper.Builder) {
+                throw new IllegalArgumentException("@timestamp must be [date] or [date_nanos]");
+            }
```

Comment:
Data streams also automatically `@timestamp` as a `date` but the way that do it seems really heavy to me. They have a special template called `DataStreamTemplate` and they automatically apply it if the name starts with `.ds-`. We could do something similar but it'd take a bunch of extra layers of plumbing to get the setting into the template resolution layer. We'd still want the validation *somewhere* anyway.

Reply:
It depends, do you want the validation outside of data streams, or as a configuration option on the data stream itself? If you want it outside of data streams (for use with aliases for instance), then don't use the `DataStreamTemplate`.

Also, I notice that you add the mapper, but don't actually enforce that the `@timestamp` is filled in the way that a data stream does. I think you also need to validate that every document contains an `@timestamp` field, is that right? Or are you planning on filling it in automatically if it is missing?

Reply:
> It depends, do you want the validation outside of data streams, or as a configuration option on the data stream itself?

I want it totally independent of data streams. If a time_series mode index is not in a data stream I still want there to be an `@timestamp` field. I was more wondering if I should make a second special template or use a mechanism like this one here.

> Also, I notice that you add the mapper, but don't actually enforce that the `@timestamp` is filled in the way that a data stream does. I think you also need to validate that every document contains an `@timestamp` field, is that right? Or are you planning on filling it in automatically if it is missing?

I was planning on adding that test in a follow up change.

Reply:
This approach seems fine to me then, since you want it to be independent.

> I was planning on adding that test in a follow up change.

This should hopefully be easy as using `DataStreamTimestampFieldMapper` instead of the `DateFieldMapper`. I think we should probably unify them also so that they share the same infrastructure for validation, for example, it doesn't help if we add a `@timestamp` field but the user configures it for `indexed: false` so we can't use it. The `DataStreamTimestampFieldMapper` does more validation here.

Reply:
I can have a look at unifying them in the follow up, yeah!